### ISSUE TITLE
Missing source code in 1.12.4 ⇒ not redistributable under normal OSS rules

### ISSUE THREAD

**Issue by mirabilos** [jquery/jquery#4445]
I’m currently trying to upgrade the jQuery embedded in an older project from 1.10.0 to 1.12.4 and found multiple horrors, called the new `external/` subdirectory.

jQuery 1.12.4 ships a minified-only npo v0.7.8-a with no source available, which makes jQuery 1.12.4 not Open Source, and not suitable for Debian to boot.

The npo repository does not have a git tag corresponding to v0.7.8-a (see https://github.com/getify/native-promise-only/issues/57 for the bug for that) and does not contain the minified compilation output, so I cannot find out the correct source code myself.

This makes jQuery 1.12.4 non-redistributable under OSS rules.

**Comment by timmywil** [comment]
Thanks for opening an issue. npo is just a copy of the module from npm, rather than reading from the `node_modules` directory. It is only used for testing. Also, minified source is still source, but you can find the unminified source code on [GitHub](https://github.com/kapersoft/npo) and [npm](https://www.npmjs.com/package/npo).

**Comment by getify** [comment]
@timmywil for clarification... you linked to a repo and package called "npo", which are a CLI utility for watching quicktime streams. I am doubting that's what is shipped with jquery, but just want to double check. Is it "native-promise-only" (that I typically abbreviate "npo") that you ship, or actually this quicktime utility "npo"?

The native-promise-only repo and package are here:

* https://github.com/getify/native-promise-only
* https://npmjs.org/package/native-promise-only

**Comment by timmywil** [comment]
Sorry, I just did `npm home npo` instead of `npm home native-promise-only` and didn't look that closely. Your links are the right ones.

**Comment by timmywil** [comment]
Fixed the links. Thanks!

**Comment by timmywil** [comment]
BTW, we do copy over license files as well when they are available in the npm packages, but they weren't available in native-promise-only or require-js. But the files copied do include links to their licenses in the comment banners. Also, this is something that hasn't been touched in 4 years.

**Comment by mirabilos** [comment]
Yes, you do copy over the files and, if present, licence statements, which fulfils the licence itself.

However, for reproducibility, if anything minified (think: object file, compiled code) is included, the source for that should also be included, so that people who use your stuff know exactly which source the thing is built from, can rebuild it if desired (consider: object file containing trojan which is not present in the source, or (which I’ve seen happen) person forgot to rebuild the binary for a release, so you actually embed the older without noticing, and if someone recompiles suddenly things stop working) or necessary (e.g. patching its source code).

For distributions like Debian, source code has to always be included. We strip binaries (`.o` and `.exe` files or, like here, minified Javascript) from the source tarballs normally and replace them by the versions compiled from the relevant package.

**Comment by mirabilos** [comment]
Incidentally, I probably found the correct commit of npo to tag, but I cannot reproduce the exact binary shipped by jQuery, probably due to a difference in minifier versions.

Another good reason why one ought to always include source…

**Comment by timmywil** [comment]
I understand it's harder to read minified source, but all the source is still available. It's not a binary. It's just minified JS code. If all you're suggesting is that we include the unminified source instead, I'm fine with that, but there are prettification tools that make minified JS easier to read.

The unminified source is included by reference indirectly through the package.json which lists `native-promise-only` as a dev dependency. You're right that the tag is not available on the Github repo anymore, but it is still available on npm. You can see everything that was published there and get the exact unminified source before the minification if that's all you need.

As far as necessary patching, if there was an issue in npo, you could do a PR on GitHub, and then we could then theoretically update the jQuery package to include the new version. That would be the better way to handle things so that everyone could benefit from the patch.

**Comment by timmywil** [comment]
I've opened a PR to address the readability issue, but this more of a mole hill than a mountain. https://github.com/jquery/jquery/pull/4446

**Comment by mirabilos** [comment]
Thanks. Another case when you’d want Complete Corresponding Source, recursive into all dependencies, is when you use jQuery in a, for example, GPL or AGPL project.

This affects all of them (qunit, sizzle, …) and their dependencies as well…

In my current project I’m subtree-merging all the sources for all the dependencies shipped, recursively, at the moment, which is making me a little crazy. (I dropped `dist/` from requirejs though, as neither the fonts nor the markdown implementation are needed to compile anything shipped, which is the condition needed here.)

**Comment by mgol** [comment]
@timmywil The `native-promise-only` package is only present on `master` & on the `3.4-stable` branch; it's not present on `1.12-stable` or `2.2-stable`.

@mirabilos We're not maintaining 1.12 & 2.2 lines anymore so I'm afraid your issue won't be resolved in those branches. However, note that `native-promise-only` is only used in tests (it doesn't influence the final jQuery source) and only to provide a polyfill for native `Promise`. If you remove the file from your copy of the branch, tests should still run in all browsers that implement the Promise interface. You can also swap the file for the latest copy of `native-promise-only` unminified source as @timmywil did in #4446 for `master`. Hopefully that's enough to resolve your issues.

**Comment by mirabilos** [comment]
Cześć Michał,

>@mirabilos We're not maintaining 1.12 & 2.2 lines anymore so I'm
>@afraid your issue won't be resolved in those branches. However, note

OK.

>@that `native-promise-only` is only used in tests (it doesn't
>@influence the final jQuery source) and only to provide a polyfill for

In that case I could just remove it indeed.

>@`master`. Hopefully that's enough to resolve your issues.

Should be, yes. But maybe you can take it under advice to ship
Complete Corresponding Souce for anything you actually distribute?
I don’t know if you have a good way to reference such dependencies
external instead of bundling them, but perhaps long-term?

bye,
//mirabilos
-- 
(gnutls can also be used, but if you are compiling lynx for your own use,
there is no reason to consider using that package)
	-- Thomas E. Dickey on the Lynx mailing list, about OpenSSL

**Comment by mgol** [comment]
@mirabilos 
> But maybe you can take it under advice to ship
> Complete Corresponding Souce for anything you actually distribute?

We can make sure to include source for things we keep in our repository. However, we have limited control over what's in `node_modules` after calling `npm install`. The dependency tree for most projects is quite large and many packages don't share full unminified code there, and for a good reason - such code would often be a big overhead and it adds up if many packages did it. Dependencies usually specify their licenses & point to their repository but that doesn't mean it's always easy to achieve exactly the same minified code from the source if the package repository doesn't maintain a lockfile.

**Comment by mirabilos** [comment]
Michał Gołębiowski-Owczarek dixit:

>We can make sure to include source for things we keep in our
>repository.

You really ought to. Or not include them.

>However, we have limited control over what's in `node_modules` after
>calling `npm install`. The dependency tree for most projects is quite
>large and many packages don't share full unminified code there, and
>for a good reason - such code would often be a big overhead and it
>adds up if many packages did it. Dependencies usually specify their
>licenses & point to their repository but that doesn't mean it's always
>easy to achieve exactly the same minified code from the source
[ lockfile excluded as it’s difficult anyway ]

Yeah, I just did basically that, figure out what all went into
jQuery 1.12.4, since I need to do the full licence analysis on
what *we* ship. It’s “not fun”, recursively.

**Comment by mgol** [comment]
@mirabilos Is there any reason you need to include anything apart from the built jQuery file (full & slim builds)? Optionally, if you want, you can also include the `src` directory that contains our AMD modules if you need that. If you do that, you'll also need `external/sizzle/dist/sizzle.js` because of [src/selector-sizzle.js](https://github.com/jquery/jquery/blob/1.12.4/src/selector-sizzle.js) but nothing else from `external` should be necessary.

**Comment by dmethvin** [comment]
The bits in any `devDependency` are not shipped by a client/user of jQuery so their licenses should not matter. We may have used Adobe Photoshop to create some of the images used for testing but that does not mean you need a Photoshop license or the source code for it. The files in `external` are only in the repo because we had no other choice but to manage our development tool dependencies manually before the days of `npm` and lockfiles.

**Comment by mgol** [comment]
@dmethvin That made me think that our current setup is unnecessarily complex & we don't need the external directory anymore. A removal PR: https://github.com/jquery/jquery/pull/4466.

**Comment by mirabilos** [comment]
On Fri, 16 Aug 2019, Michał Gołębiowski-Owczarek wrote:

> @mirabilos Is there any reason you need to include anything apart from
> @the built jQuery file (full & slim builds)? Optionally, if you want,

Yes, we want to ship complete corresponding source for everything
we ship. In some cases, this is even a requirement of the larger
project licence (something GPL-ish), but even if not (such as in
the current project, where we managed to get rid of all GPL stuff)
it’s good practice (also for other things like in Debian).

> @you can also include the `src` directory that contains our AMD

By default, I included your git repository, then looked at what
was missing from there. I probably could have removed parts of
external/ first but I was unsure which was really needed for
the build and which was not, considering…

> @modules if you need that. If you do that, you'll also need
> @`external/sizzle/dist/sizzle.js` because of
> @[src/selector-sizzle.js](https://github.com/jquery/jquery/blob/1.12.4/src/selector-sizzle.js)
> @but nothing else from `external` should be necessary.

… but yeah.

In the end, I had to rebuild jQuery locally anyway, backporting
a CVE fix and reverting the removal of another CVE fix, so I had
to include some of the stuff and reverse-engineer one of your
buildsystems until my output about matched what you did (since
I refuse to use npm or any other system-polluting “package” ma‐
nager).

I succeeded, but it wasn’t fun.

bye,
//mirabilos
-- 
tarent solutions GmbH
Rochusstraße 2-4, D-53123 Bonn • http://www.tarent.de/
Tel: +49 228 54881-393 • Fax: +49 228 54881-235
HRB 5168 (AG Bonn) • USt-ID (VAT): DE122264941
Geschäftsführer: Dr. Stefan Barth, Kai Ebenrett, Boris Esser, Alexander Steeg

**********

Mit der tarent Academy bieten wir auch Trainings und Schulungen in den
Bereichen Softwareentwicklung, Agiles Arbeiten und Zukunftstechnologien an.

Besuchen Sie uns auf www.tarent.de/academy. Wir freuen uns auf Ihren Kontakt.

**********
